// @hibas123/logging/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logging = exports.TerminalFormats = exports.FormatTypes = exports.Colors = exports.DefaultColorMap = exports.LoggingTypes = exports.withColor = exports.removeColors = exports.LoggingBase = exports.ConsoleAdapter = void 0;
const base_js_1 = require("./base.js");
var consolewriter_js_1 = require("./consolewriter.js");
Object.defineProperty(exports, "ConsoleAdapter", { enumerable: true, get: function () { return consolewriter_js_1.ConsoleAdapter; } });
var base_js_2 = require("./base.js");
Object.defineProperty(exports, "LoggingBase", { enumerable: true, get: function () { return base_js_2.LoggingBase; } });
Object.defineProperty(exports, "removeColors", { enumerable: true, get: function () { return base_js_2.removeColors; } });
Object.defineProperty(exports, "withColor", { enumerable: true, get: function () { return base_js_2.withColor; } });
var types_js_1 = require("./types.js");
Object.defineProperty(exports, "LoggingTypes", { enumerable: true, get: function () { return types_js_1.LoggingTypes; } });
Object.defineProperty(exports, "DefaultColorMap", { enumerable: true, get: function () { return types_js_1.DefaultFormatConfig; } });
Object.defineProperty(exports, "Colors", { enumerable: true, get: function () { return types_js_1.Colors; } });
Object.defineProperty(exports, "FormatTypes", { enumerable: true, get: function () { return types_js_1.FormatTypes; } });
Object.defineProperty(exports, "TerminalFormats", { enumerable: true, get: function () { return types_js_1.TerminalFormats; } });
exports.Logging = undefined;
exports.Logging = new base_js_1.LoggingBase();
exports.default = exports.Logging;
