// @hibas123/logging/esm/types.js
export var LoggingTypes;
(function (LoggingTypes) {
    LoggingTypes[LoggingTypes["Debug"] = 0] = "Debug";
    LoggingTypes[LoggingTypes["Log"] = 1] = "Log";
    LoggingTypes[LoggingTypes["Warning"] = 2] = "Warning";
    LoggingTypes[LoggingTypes["Error"] = 3] = "Error";
})(LoggingTypes || (LoggingTypes = {}));
export const TerminalFormats = {
    Reset: "\x1b[0m",
    Bold: "\x1b[1m",
    Underscore: "\x1b[4m",
    Blink: "\x1b[5m",
    Reverse: "\x1b[7m",
    Hidden: "\x1b[8m",
    FgBlack: "\x1b[30m",
    FgRed: "\x1b[31m",
    FgGreen: "\x1b[32m",
    FgYellow: "\x1b[33m",
    FgBlue: "\x1b[34m",
    FgMagenta: "\x1b[35m",
    FgCyan: "\x1b[36m",
    FgWhite: "\x1b[37m",
    BgBlack: "\x1b[40m",
    BgRed: "\x1b[41m",
    BgGreen: "\x1b[42m",
    BgYellow: "\x1b[43m",
    BgBlue: "\x1b[44m",
    BgMagenta: "\x1b[45m",
    BgCyan: "\x1b[46m",
    BgWhite: "\x1b[47m",
};
export var FormatTypes;
(function (FormatTypes) {
    FormatTypes[FormatTypes["COLOR"] = 0] = "COLOR";
    FormatTypes[FormatTypes["BOLD"] = 1] = "BOLD";
    FormatTypes[FormatTypes["UNDERSCORE"] = 2] = "UNDERSCORE";
    FormatTypes[FormatTypes["BLINK"] = 3] = "BLINK";
})(FormatTypes || (FormatTypes = {}));
export var Colors;
(function (Colors) {
    Colors[Colors["NONE"] = 0] = "NONE";
    Colors[Colors["RED"] = 1] = "RED";
    Colors[Colors["GREEN"] = 2] = "GREEN";
    Colors[Colors["YELLOW"] = 3] = "YELLOW";
    Colors[Colors["BLUE"] = 4] = "BLUE";
    Colors[Colors["MAGENTA"] = 5] = "MAGENTA";
    Colors[Colors["CYAN"] = 6] = "CYAN";
    Colors[Colors["WHITE"] = 7] = "WHITE";
})(Colors || (Colors = {}));
function colorFormat(color) {
    return {
        type: FormatTypes.COLOR,
        color,
    };
}
const boldFormat = {
    type: FormatTypes.BOLD,
};
export class DefaultFormatConfig {
    constructor() {
        this.error = [colorFormat(Colors.RED), boldFormat];
        this.warning = [colorFormat(Colors.YELLOW), boldFormat];
        this.log = [colorFormat(Colors.NONE), boldFormat];
        this.debug = [colorFormat(Colors.CYAN), boldFormat];
        this.date = [colorFormat(Colors.NONE)];
        this.file = [colorFormat(Colors.NONE)];
    }
}
